src/nvim/api/vim.c

```diff
@@ -780,7 +797,40 @@ void nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history, Dict(
     kind = opts->err ? "echoerr" : history ? "echomsg" : "echo";
   }
 
-  msg_multihl(hl_msg, kind, history, opts->err);
+  bool is_kind_progress = kind != NULL && strcmp(kind, "progress") == 0;
+
+  if (!is_kind_progress && (opts->status.size != 0 || opts->title.size != 0 || opts->percent > 0)) {
+    api_set_error(err, kErrorTypeValidation,
+                  "title, status and percents fields can only be used with progress messages");
```

Comment:
use one of the VALIDATE macros. same for most other `api_set_error` cases below, except in rare cases.